lib/TolerantVariableReferenceFinder.php

```diff
@@ -0,0 +1,180 @@
+            false === $node instanceof UseVariableName &&
+            false === $node instanceof Parameter
+        ) {
+            return null;
+        }
+
+        if (
+            ($node instanceof Variable && $node->parent instanceof ScopedPropertyAccessExpression)
+            || ($node instanceof Variable && $node->getFirstAncestor(PropertyDeclaration::class))
+        ) {
+            return null;
+        }
+
+        return $node;
+    }
+
+    private function scopeNode(Node $variable): Node
+    {
+        $name = $this->variableName($variable);
+        if ($variable instanceof UseVariableName) {
+            $variable = $variable->getFirstAncestor(MethodDeclaration::class) ?: $variable;
+        }
+
+        $scopeNode = $variable->getFirstAncestor(FunctionLike::class, ClassLike::class, SourceFileNode::class);
+        while (
+            $scopeNode instanceof AnonymousFunctionCreationExpression &&
+            $this->nameExistsInUseClause($name, $scopeNode)
+        ) {
+            $scopeNode = $scopeNode->getFirstAncestor(FunctionLike::class, ClassLike::class, SourceFileNode::class);
+        }
+
+        if (null === $scopeNode) {
+            throw new \Exception(
+                'Could not determine scope node, this should not happen as ' .
+                'there should always be a SourceFileNode.'
+            );
+        }
+
+        return $scopeNode;
+    }
+
+    private function nameExistsInUseClause(string $variableName, AnonymousFunctionCreationExpression $function): bool
+    {
+        if (
+            $function->anonymousFunctionUseClause === null
+            || $function->anonymousFunctionUseClause->useVariableNameList === null
+            || $function->anonymousFunctionUseClause->useVariableNameList instanceof MissingToken
+        ) {
+            return false;
+        }
+
+        foreach ($function->anonymousFunctionUseClause->useVariableNameList->getElements() as $useVariableName) {
+            assert($useVariableName instanceof UseVariableName);
+            if ($this->variableName($useVariableName) == $variableName) {
+                return true;
+            }
+        }
+        return false;
+    }
+    /**
+     * @return Generator<PotentialLocation>
+     */
+    private function find(Node $scopeNode, string $name, string $uri): Generator
+    {
+        /** @var Node $node */
+        foreach ($scopeNode->getChildNodes() as $node) {
+            if ($node instanceof AnonymousFunctionCreationExpression && !$this->nameExistsInUseClause($name, $node)) {
+                continue;
+            }
+            
+            if (
+                $this->isPotentialReferenceNode($node)
+                && $name == $this->variableName($node)
+            ) {
+                if (!$this->foundDefinition) {
+                    // We artificially skip the first result
+                    // assuming it's the definition
+                    $this->foundDefinition = true;
```

Comment:
What is the motivation here ?
I would like to have the definition as a result too personally, especially since in PHP people tends to not declare their variables until they need to affect a value to it.
But it might be only me ;)

If it still make sense to you to ignore it, it might be better to remove the attribute `foundDefinition` from the class and make it a parameter of the method.
This notion should not get out of this method and currently it forces you to know the behavior in order to be aware that you need to set it to `true` before calling this method for the first time.
It's kind of licking everywhere I mean, not sure my explanation is that good though...

Reply:
The motivation was that all other reference finders did not return the definition. Probably because the definition is not a reference, not sure. The definition locator finds the first occurrence of variables as well. If I left that here then it is that location is seen twice in "Find all references".

Reply:
OK, for me a reference is just a place where the variable is used.
I don't make a difference between declaration, assignation, usage, etc.

> The definition locator finds the first occurrence of variables as well

That's its purpose, I guess you meant that it uses the same logic: the first time we see the variable is the definition
I've nothing against this idea, since it's impossible to manually declare a variable in PHP it make perfect sense.

What do you call the "Find all references" exactly ? Which handler ?
If it's shown twice it would suggest that the logic in this feature is to separate the definition of the reference indeed.
But then when does this `TolerantVariableReferenceFinder` would be used ?

Basically I just tried to make sure that when we ask for the references of a variables we have all the results, if it's aggregated with a definition locator great then.
Like currently if I leave my cursor on a variable it hovers all the occurrences, including the declaration.

Reply:
> What do you call the "Find all references" exactly ? Which handler ?

`Phpactor\Extension\LanguageServerReferenceFinder\Handler\ReferencesHandler`

> But then when does this TolerantVariableReferenceFinder would be used ?

What do you mean? My point is that I just followed the way every other `ReferenceFinder` works - to not return the definition. I am fine with either way, but it must be consistent between all the `ReferenceFinder`s, I think.

>Basically I just tried to make sure that when we ask for the references of a variables, we have all the results, if it's aggregated with a definition locator great then.

I agree. I took the code form the `ReferencesHandler`, because to rename something you are basically renaming a subset of the references, but you need to check them all. There the references were combined with the defintion, so I went along. I don't mind reworking my code. I agree with your logic, but I had to be consistent with the existing code.

> Like currently if I leave my cursor on a variable it hovers all the occurrences, including the declaration.

I don't think the Hover handler uses references at all. It just highlight the text which includes the variables with the same name in other methods.

Reply:
Thanks for the handler, make more sense to me know, depending on the context we will add the definition or not.
So you're right to keep it like that, I was worried that if we tried to implement the rename with only the finder we would endup not renaming the definition.

But, for me who likes to see the definition it will work because I will configure the context accordingly.
For the rename we will have to be aware that will also need to retrieve the definition.

What about the suggestion of removing the class attribute and pass it as arguments to the method ?

Reply:
Actually I liked your suggestion and just tried it. It didn't work. I didn't have enough time to read, but would seem that you can't change a variable within the `foreach`, because the yields seem to be executed in some different order or something. The problem was the recursive call - when the definition was found, the `true` value was used only for _that_ cycle, but not the next cycles, if that makes sense.
If you think I implemented it wrong, I can check in the non-working code so you can take a look.

Reply:
Something like:
```php
    /**
     * @return Generator<PotentialLocation>
     */
    private function find(Node $scopeNode, string $name, string $uri, bool $foundDefinition = false): Generator
    {
        /** @var Node $node */
        foreach ($scopeNode->getChildNodes() as $node) {
            if ($node instanceof AnonymousFunctionCreationExpression && !$this->nameExistsInUseClause($name, $node)) {
                continue;
            }
            
            if (
                $this->isPotentialReferenceNode($node)
                && $name == $this->variableName($node)
            ) {
                if (!$foundDefinition) {
                    // We artificially skip the first result
                    // assuming it's the definition
                    $foundDefinition = true;
                } else {
                    yield PotentialLocation::surely(
                        Location::fromPathAndOffset($uri, $node->getStart())
                    );
                }
            } else {
                yield from $this->find($node, $name, $uri, $foundDefinition);
            }
        }
    }
```

Reply:
That is almost exactly what I tried. If I had to guess, at the time `yield from $this->find($node, $name, $uri, $foundDefinition);` was executed, the value of `$foundDefinition` is already different. When I did some dumping I saw that after the iteration where `$foundDefinition` became true, it was again false for the next iteration. Not sure what am I doing wrong.
Will try again and post the results.

Reply:
Boy am I dumb. This doesn't work because this is a recursion rather than a flat `foreach`-ing. If a call stack set the argument to `true`, we want all other branches to know about it. Look at this call stack:
```
.^ "Found: "
^ "Found: "
^ "  Found: "
^ "    Found: "
^ "    Found: 1" <------ Here it finds the definition
^ "Found: " <---- We now want this to know about it as well
^ "  Found: "
^ "    Found: "
^ "    Found: "
^ "      Found: "
^ "      Found: 1"
```
This means we should pass it by reference, which I am not terribly fond of.

Reply:
Using reference is OK for me, it's not a bad thing as long as it is justified.
When you think about it, before you're using a class attribute which have a global scope. This is not really different, if you're uncomfortable with it because you think it's unclear you can always document it in the phpdoc to explain it a bit more.

Reply:
Instead of doing that, why not just discard the first result?

Reply:
Indeed, I didn't find any indication in the documentation that would suggest `yield from` does not keep the ordering.
The find method could then be used to return all references, included the definition.

Reply:
Who should discard it? And how?

Reply:
Delegate to anoter method:

```
$generator = $this->doFindReference($blah);
$generator->next();
yield from $generator;
```

or similar, probably a better way :smile:

Reply:
> Using reference is OK for me, it's not a bad thing as long as it is justified.
> When you think about it, before you're using a class attribute which have a global scope. This is not really different, if you're uncomfortable with it because you think it's unclear you can always document it in the phpdoc to explain it a bit more.

I think it would make the code less readable, yes. And it pollutes all the calls with the initialisation of the variable, especially the first one where it is not relevant, because it is internal to the method. It is, in fact, mimicking an instance variable for a smaller scope.

Reply:
> $generator = $this->doFindReference($blah);

I actually like that. Making the change.

Reply:
Pushed.